File: src/main/java/com/example/reactive/chapter2/sse/TemperatureController.java
package com.example.reactive.chapter2.sse;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

import lombok.extern.slf4j.Slf4j;

@Slf4j
@Controller
public class TemperatureController {

    private List<SseEmitter> emitters = new ArrayList<>();

    @GetMapping("temperature-stream")
    public SseEmitter temperatureStream() {

        SseEmitter emitter = new SseEmitter(60000L);
        emitter.onCompletion(() -> emitters.remove(emitter));
        emitter.onTimeout(() -> emitters.remove(emitter));
        emitters.add(emitter);

        return emitter;
    }

    @Async
    @EventListener
    public void onEvent(Temperature temperature) {

        List<SseEmitter> deadEmitters = new ArrayList<>();

        log.info("[{}] Event received : {}", Thread.currentThread().getName(), temperature.getValue());

        emitters.forEach(emitter -> {
            try {
                emitter.send(temperature);

            } catch (IOException e) {
                deadEmitters.add(emitter);
            }
        });

        emitters.removeAll(deadEmitters);
    }
}

File: src/main/kotlin/wmp/edu/reactive/temperature/flux/TemperatureController.kt
package wmp.edu.reactive.temperature.flux

import org.springframework.http.codec.ServerSentEvent
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import wmp.edu.reactive.shared.logger

@RestController
class TemperatureController(val sensor: TemperatureSensor) {
    @GetMapping("/temperature-stream")
    fun prove(): Flux<ServerSentEvent<Temperature>> {
        return sensor.prove()
            .map {
                logger().info("event received! : $it")
                ServerSentEvent.builder(it).build()
            }
    }
}
